makefiles/dependency_resolution.inc.mk

```diff
@@ -0,0 +1,29 @@
+# Perform a recursive dependency resolution: Include $(RIOTBASE)/Makefile.dep
+# until no new modules and pkgs are pull in order to catch all transient
+# dependencies
+
+OLD_USEMODULE := $(sort $(USEMODULE))
+OLD_USEPKG:= $(sort $(USEPKG))
+
+# pull in dependencies of the currently used modules and pkgs
+include $(RIOTBASE)/Makefile.dep
+
+# check if required features are provided and update $(FEATURES_USED)
+include $(RIOTMAKE)/features_check.inc.mk
+
+# translate used features into used module, where needed
+include $(RIOTMAKE)/features_modules.inc.mk
+
+# sort and de-duplicate used modules and pkgs
+USEMODULE := $(sort $(USEMODULE))
+USEPKG := $(sort $(USEPKG))
+
+# If set of used modules and pkgs has changed during last run, run again to
+# recursively catch transitive dependencies
+ifneq ($(OLD_USEMODULE) $(OLD_USEPKG),$(USEMODULE) $(USEPKG))
+  include $(RIOTMAKE)/dependency_resolution.inc.mk
+endif
+
+# add default modules again, as $(DEFAULT_MODULE) might have been extended
+# during dependency processing
+USEMODULE += $(filter-out $(DISABLE_MODULE),$(DEFAULT_MODULE))
```

Comment:
The way this is done how changes the old behavior. Disabling `auto-init` would no longer disable all `auto-init_%` modules.

Reply:
That seems to also be not the case in `master`:

```
DISABLE_MODULE=auto_init BOARD=arduino-due make info-modules -C examples/default/
make: Entering directory '/home/maribu/Repos/software/RIOT/examples/default'
auto_init_saul
board
boards_common_arduino_due
core
core_init
core_msg
core_panic
cortexm_common
cortexm_common_periph
cpu
fmt
isrpipe
newlib
newlib_nano
newlib_syscalls_default
periph
periph_common
periph_gpio
periph_init
periph_init_common
periph_init_gpio
periph_init_init
periph_init_pm
periph_init_uart
periph_pm
periph_uart
phydat
ps
saul
saul_default
saul_gpio
saul_init_devs
saul_reg
shell
shell_commands
stdin
stdio_uart
stdio_uart_rx
sys
tsrb
make: Leaving directory '/home/maribu/Repos/software/RIOT/examples/default'
```

Reply:
Oh, well I stand corrected the. It does work poorly since it depends on order of inclusion and therefore some of `auto_int_%` modules get added straight away.

Reply:
Ok so the behavior is broken anyway, this is not good but I can't expect this PR to handle something that wasn't handled before. I think this can be fixed by changing the variable name not `DEFAULT_MODULE` but `DEFAULT_MODULE_OPTIONAL` or something like that , https://github.com/RIOT-OS/RIOT/issues/13651#issuecomment-604525736_. 

Anyway lets keep this as is then, I'll try to cleanup in a follow-up.

Reply:
Thanks for tackling this in a follow up!

Reply:
tackled in #13785